websocketOutportDriver/sender/webSocketSender.go

```diff
@@ -125,148 +114,119 @@ func (w *webSocketSender) handleReceiveAck(client *webSocketClient) {
 			continue
 		}
 
-		w.mutAcknowledges.Lock()
-		w.acknowledges[client.remoteAddr][counter] = struct{}{}
-		w.mutAcknowledges.Unlock()
+		w.acknowledges.AddReceivedAcknowledge(client.remoteAddr, counter)
 	}
 }
 
 func (w *webSocketSender) start() {
 	err := w.server.ListenAndServe()
-	if err != nil {
+	if err != nil && !strings.Contains(err.Error(), "http: Server closed") {
 		w.log.Error("could not initialize webserver", "error", err)
 	}
 }
 
-func (w *webSocketSender) sendWithRetrial(data []byte, ackData []byte, counter uint64) {
-	clients := w.getClients()
+func (w *webSocketSender) sendDataToClients(
+	data []byte,
+	counter uint64,
+) error {
+	numSent := 0
+	var err error
+
+	clients := w.clientsHolder.GetAll()
 	if len(clients) == 0 {
-		w.log.Warn("no client to send to")
+		return ErrNoClientToSendTo
 	}
 
-	ticker := time.NewTicker(time.Millisecond * retrialTimeoutMS)
-	defer ticker.Stop()
-
-	for {
-		select {
-		case <-ticker.C:
-			clients = w.getClients()
-			if len(clients) == 0 {
-				continue
-			}
-			dataSent := w.sendDataToClients(data, ackData, clients, counter)
-			if dataSent {
-				return
-			}
+	for _, client := range w.clientsHolder.GetAll() {
+		err = w.sendData(data, *client, counter)
+		if err != nil {
+			w.log.Error("couldn't send data to client", "error", err)
+			continue
 		}
-	}
-}
 
-func (w *webSocketSender) sendDataToClients(
-	data []byte,
-	ackData []byte,
-	clients map[string]*webSocketClient,
-	counter uint64,
-) bool {
+		numSent++
+	}
 
-	for _, client := range clients {
-		result := w.sendData(data, ackData, *client, counter)
-		if !result {
-			return false
-		}
+	if numSent == 0 {
+		return fmt.Errorf("data wasn't sent to any client. last known error: %w", err)
 	}
 
-	return true
+	return nil
 }
 
 func (w *webSocketSender) sendData(
 	data []byte,
-	ackData []byte,
 	client webSocketClient,
 	counter uint64,
-) bool {
+) error {
+	if len(data) == 0 {
+		return ErrEmptyDataToSend
+	}
+
 	errSend := client.conn.WriteMessage(websocket.BinaryMessage, data)
 	if errSend != nil {
 		w.log.Warn("could not send data to client", "remote addr", client.remoteAddr, "error", errSend)
-		return false
-	}
-
-	if len(data) == 0 || len(ackData) == 0 {
-		return false
+		return fmt.Errorf("%w while writing message to client %s", errSend, client.remoteAddr)
 	}
 
-	if ackData[0] == prefixWithoutAck[0] {
-		return true
+	if !w.withAcknowledge {
+		return nil
 	}
 
 	// TODO: might refactor this (send to each clients, then wait for all VS send to one client, wait for it, move to next)
 	w.waitForAck(client.remoteAddr, counter)
 
-	return true
+	return nil
 }
 
 func (w *webSocketSender) waitForAck(remoteAddr string, counter uint64) {
 	for {
-		acknowledges := w.getAcknowledges()
-		acksForAddress, ok := acknowledges[remoteAddr]
+		acksForAddress, ok := w.acknowledges.GetAcknowledgesOfAddress(remoteAddr)
```

Comment:
Why call it `GetAcknowledgesOfAddress` when you are only waiting for **one** acknowlege from a single client?

Reply:
no, it is an instance of `websocketClientAcknowledgesHolder` that contains a map underneath